File: packages/updater-core/ios/Sources/ArtifactInstaller.swift
import CryptoKit
import Foundation
import ZIPFoundation

public struct RNFileIdentity {
  public let path: String
  public let hash: String
  public let size: Int64
}

public final class ArtifactInstaller {
  private let folding: [String: String]
  private let maxBytes: Int64 = 100 * 1024 * 1024

  public init(caseFoldingData: Data) throws {
    folding = try JSONDecoder().decode([String: String].self, from: caseFoldingData)
  }

  public func validatePaths(_ paths: [String]) throws {
    var nodes: [String: (String, Bool)] = [:]
    guard paths.count <= 5000 else { throw LaunchError.incompatibleArtifact }
    for path in paths {
      guard
        !path.isEmpty && path.utf8.count <= 512 && !path.contains("\\")
          && !path.unicodeScalars.contains(where: { $0.value < 32 || $0.value == 127 })
      else { throw LaunchError.incompatibleArtifact }
      let components = path.components(separatedBy: "/")
      var spelling = ""
      for (index, part) in components.enumerated() {
        guard !part.isEmpty && part != "." && part != ".." && part.utf8.count <= 255 else {
          throw LaunchError.incompatibleArtifact
        }
        spelling += (index == 0 ? "" : "/") + part
        let folded = spelling.decomposedStringWithCanonicalMapping.unicodeScalars.map {
          folding[String($0)] ?? String($0)
        }.joined().decomposedStringWithCanonicalMapping
        guard
          !(index == 0
            && ["bundle.json", "otakit_files.json", "otakit-embedded.json"].contains(folded))
        else { throw LaunchError.incompatibleArtifact }
        let directory = index < components.count - 1
        if let prior = nodes[folded] {
          guard Array(prior.0.utf8) == Array(spelling.utf8) && prior.1 == directory && directory
          else { throw LaunchError.incompatibleArtifact }
        }
        nodes[folded] = (spelling, directory)
      }
    }
    guard paths.contains("index.bundle") && paths.contains("otakit-bundle.json") else {
      throw LaunchError.incompatibleArtifact
    }
  }

  public func installZip(
    archiveURL: URL, into root: URL, manifest: RNManifest, bundleKeys: [String: Data],
    rnVersion: String, bytecodeVersion: UInt32
  ) throws -> [RNFileIdentity] {
    guard let archiveSize = try archiveURL.resourceValues(forKeys: [.fileSizeKey]).fileSize,
      archiveSize == manifest.size
    else { throw LaunchError.incompatibleArtifact }
    var bytes = try Data(contentsOf: archiveURL)
    guard bytes.count == manifest.size && Self.hash(bytes) == manifest.sha256 else {
      throw LaunchError.incompatibleArtifact
    }
    if let encryption = manifest.encryption {
      guard let key = bundleKeys[encryption.kid], key.count == 32,
        String(Self.hash(key).prefix(16)) == encryption.kid,
        let wrapNonce = Data(base64Encoded: encryption.wrapNonce),
        let wrapped = Data(base64Encoded: encryption.wrappedDek),
        let nonce = Data(base64Encoded: encryption.nonce)
      else { throw LaunchError.incompatibleArtifact }
      let dek = try AES.GCM.open(
        AES.GCM.SealedBox(
          nonce: AES.GCM.Nonce(data: wrapNonce), ciphertext: wrapped.prefix(32),
          tag: wrapped.suffix(16)), using: SymmetricKey(data: key))
      guard bytes.count >= 16 else { throw LaunchError.incompatibleArtifact }
      bytes = try AES.GCM.open(
        AES.GCM.SealedBox(
          nonce: AES.GCM.Nonce(data: nonce), ciphertext: bytes.dropLast(16), tag: bytes.suffix(16)),
        using: SymmetricKey(data: dek))
    }
    let archive = try Archive(data: bytes, accessMode: .read)
    let entries = Array(archive)
    // RN exporter emits files only. Reject link/alias/extra directory entries before extraction.
    guard entries.count <= 5000 else { throw LaunchError.incompatibleArtifact }
    var extractedSize: UInt64 = 0
    for entry in entries {
      guard entry.type == .file && entry.uncompressedSize <= UInt64(maxBytes) - extractedSize else {
        throw LaunchError.incompatibleArtifact
      }
      extractedSize += entry.uncompressedSize
    }
    try validatePaths(entries.map(\.path))
    // The caller owns a newly-created empty root. Resolve its filesystem aliases once.
    let canonicalRoot = root.resolvingSymlinksInPath()
    guard try FileManager.default.contentsOfDirectory(atPath: canonicalRoot.path).isEmpty else {
      throw LaunchError.incompatibleArtifact
    }
    for entry in entries {
      let destination = canonicalRoot.appendingPathComponent(entry.path)
      guard destination.path.utf8.count + 1 <= 1024 else { throw LaunchError.incompatibleArtifact }
      try FileManager.default.createDirectory(
        at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
      _ = try archive.extract(entry, to: destination)
    }
    return try verifyDirectory(
      canonicalRoot, manifest: manifest, rnVersion: rnVersion, bytecodeVersion: bytecodeVersion,
      expectedPaths: entries.map(\.path))
  }

  public func verifyDirectory(
    _ root: URL, manifest: RNManifest, rnVersion: String, bytecodeVersion: UInt32,
    expectedPaths: [String]? = nil
  ) throws -> [RNFileIdentity] {
    let canonicalRoot = root.resolvingSymlinksInPath()
    let paths = expectedPaths ?? manifest.files?.map(\.path)
    var originalPaths: [String: String] = [:]
    if let paths = paths {
      try validatePaths(paths)
      for path in paths { originalPaths[path.decomposedStringWithCanonicalMapping] = path }
    }
    var enumerationError: Error?
    guard
      let enumerator = FileManager.default.enumerator(
        at: canonicalRoot,
        includingPropertiesForKeys: [
          .isSymbolicLinkKey, .isRegularFileKey, .isDirectoryKey, .fileSizeKey,
        ],
        errorHandler: { _, error in
          enumerationError = error
          return false
        })
    else { throw LaunchError.incompatibleArtifact }
    var files: [RNFileIdentity] = []
    var size: Int64 = 0
    for case let url as URL in enumerator {
      let properties = try url.resourceValues(forKeys: [
        .isSymbolicLinkKey, .isRegularFileKey, .isDirectoryKey, .fileSizeKey,
      ])
      guard
        properties.isSymbolicLink != true
          && (properties.isRegularFile == true || properties.isDirectory == true)
      else { throw LaunchError.incompatibleArtifact }
      if properties.isRegularFile == true {
        guard let fileSize = properties.fileSize,
          fileSize >= 0 && fileSize <= maxBytes - size && files.count < 5000
        else { throw LaunchError.incompatibleArtifact }
        size += Int64(fileSize)
        let components = url.resolvingSymlinksInPath().pathComponents
        let rootComponents = canonicalRoot.pathComponents
        guard Array(components.prefix(rootComponents.count)) == rootComponents else {
          throw LaunchError.incompatibleArtifact
        }
        let diskPath = components.dropFirst(rootComponents.count).joined(separator: "/")
        // Apple filesystems may return decomposed spelling. Hash the validated original spelling.
        let path: String
        if paths != nil {
          guard let original = originalPaths[diskPath.decomposedStringWithCanonicalMapping] else {
            throw LaunchError.incompatibleArtifact
          }
          path = original
        } else {
          path = diskPath
        }
        let data = try Data(contentsOf: url)
        guard data.count == fileSize else { throw LaunchError.incompatibleArtifact }
        files.append(RNFileIdentity(path: path, hash: Self.hash(data), size: Int64(fileSize)))
      }
    }
    if let error = enumerationError { throw error }
    if let paths = paths, files.count != paths.count { throw LaunchError.incompatibleArtifact }
    try validatePaths(files.map(\.path))
    let canonical = files.sorted {
      Array($0.path.utf8).lexicographicallyPrecedes(Array($1.path.utf8))
    }.map { "\($0.path):\($0.hash)" }.joined(separator: "\n")
    guard Self.hash(Data(canonical.utf8)) == manifest.contentHash else {
      throw LaunchError.incompatibleArtifact
    }
    if manifest.strategy == "deltas" {
      try validateDeltaInventory(manifest)
      let declared = Dictionary(uniqueKeysWithValues: manifest.files!.map { ($0.path, $0) })
      guard
        size == manifest.size && files.count == declared.count
          && files.allSatisfy({ file in
            declared[file.path]?.size == file.size && declared[file.path]?.sha256 == file.hash
          })
      else { throw LaunchError.incompatibleArtifact }
    }
    let descriptorData = try Data(
      contentsOf: canonicalRoot.appendingPathComponent("otakit-bundle.json"))
    guard descriptorData.count > 0 && descriptorData.count <= 256 * 1024,
      let descriptor = try JSONSerialization.jsonObject(with: descriptorData) as? [String: Any],
      descriptor["format"] as? String == "otakit-rn", descriptor["formatVersion"] as? Int == 1,
      descriptor["framework"] as? String == "react-native",
      descriptor["platform"] as? String == manifest.platform,
      descriptor["runtimeVersion"] as? String == manifest.runtimeVersion,
      descriptor["version"] as? String == manifest.version,
      descriptor["entryPoint"] as? String == "index.bundle",
      descriptor["engine"] as? String == "hermes",
      descriptor["bundleFormat"] as? String == "hermes-bytecode",
      descriptor["reactNativeVersion"] as? String == rnVersion
    else { throw LaunchError.incompatibleArtifact }
    let bytecode = try Data(contentsOf: canonicalRoot.appendingPathComponent("index.bundle"))
    guard
      bytecode.count >= 12
        && bytecode.prefix(8).map({ String(format: "%02x", $0) }).joined() == "c61fbc03c103191f"
    else { throw LaunchError.incompatibleArtifact }
    let version = bytecode[8..<12].enumerated().reduce(UInt32(0)) {
      $0 | UInt32($1.element) << ($1.offset * 8)
    }
    guard version == bytecodeVersion else { throw LaunchError.incompatibleArtifact }
    if descriptor.keys.contains("expo") {
      guard let expo = descriptor["expo"] as? [String: Any] else {
        throw LaunchError.incompatibleArtifact
      }
      guard expo["configFile"] as? String == "expo-config.json" else {
        throw LaunchError.incompatibleArtifact
      }
      let data = try Data(contentsOf: canonicalRoot.appendingPathComponent("expo-config.json"))
      guard data.count <= 256 * 1024 else { throw LaunchError.incompatibleArtifact }
      guard (try JSONSerialization.jsonObject(with: data)) is [String: Any] else {
        throw LaunchError.incompatibleArtifact
      }
      if let dom = expo["domRoot"] {
        guard
          dom as? String == "www.bundle"
            && files.contains(where: {
              $0.path.hasPrefix("www.bundle/") && $0.path.hasSuffix(".html")
            })
        else { throw LaunchError.incompatibleArtifact }
      }
    }
    return files
  }

  /// Delta URLs are transport hints; path/hash pairs must reproduce the signed content identity.
  public func validateDeltaInventory(_ manifest: RNManifest) throws {
    guard manifest.strategy == "deltas", let files = manifest.files else {
      throw LaunchError.incompatibleArtifact
    }
    try validatePaths(files.map(\.path))
    var total: Int64 = 0
    var sizes: [String: Int64] = [:]
    for file in files {
      guard
        file.size >= 0 && file.size <= maxBytes - total
          && file.sha256.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil
          && (sizes[file.sha256] == nil || sizes[file.sha256] == file.size)
      else { throw LaunchError.incompatibleArtifact }
      if ["index.bundle", "otakit-bundle.json"].contains(file.path) && file.size == 0 {
        throw LaunchError.incompatibleArtifact
      }
      total += file.size
      sizes[file.sha256] = file.size
    }
    let canonical = files.sorted {
      Array($0.path.utf8).lexicographicallyPrecedes(Array($1.path.utf8))
    }.map { "\($0.path):\($0.sha256)" }.joined(separator: "\n")
    guard total == manifest.size && Self.hash(Data(canonical.utf8)) == manifest.contentHash else {
      throw LaunchError.incompatibleArtifact
    }
  }

  private static func hash(_ data: Data) -> String {
    SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }
}

File: packages/updater-core/android/src/main/java/com/otakit/core/ArtifactInstaller.java
package com.otakit.core;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.security.MessageDigest;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import net.lingala.zip4j.ZipFile;
import net.lingala.zip4j.model.FileHeader;
import org.json.JSONArray;
import org.json.JSONObject;

/** RN-only installation. Validate the complete inventory before creating payload files. */
public final class ArtifactInstaller {

  public static final long MAX_BYTES = 100 * 1024 * 1024;
  private final JSONObject folding;

  public ArtifactInstaller(byte[] caseFoldingData) throws Exception {
    folding = new JSONObject(new String(caseFoldingData, StandardCharsets.UTF_8));
  }

  private String comparisonKey(String value) throws Exception {
    String normalized = Normalizer.normalize(value, Normalizer.Form.NFD);
    StringBuilder result = new StringBuilder();
    for (int offset = 0; offset < normalized.length(); ) {
      int codepoint = normalized.codePointAt(offset);
      String character = new String(Character.toChars(codepoint));
      result.append(folding.has(character) ? folding.getString(character) : character);
      offset += Character.charCount(codepoint);
    }
    return Normalizer.normalize(result, Normalizer.Form.NFD);
  }

  public void validatePaths(List<String> paths) throws Exception {
    if (paths.size() < 2 || paths.size() > 5000) throw invalid();
    Map<String, String> spellings = new HashMap<>();
    Set<String> directories = new HashSet<>();
    for (String path : paths) {
      if (
        path.isEmpty() ||
        path.getBytes(StandardCharsets.UTF_8).length > 512 ||
        path.indexOf('\\') >= 0 ||
        path.matches("(?s).*[\\x00-\\x1f\\x7f].*") ||
        !new String(path.getBytes(StandardCharsets.UTF_8), StandardCharsets.UTF_8).equals(path)
      ) throw invalid();
      String[] parts = path.split("/", -1);
      String spelling = "";
      for (int i = 0; i < parts.length; i++) {
        String part = parts[i];
        if (
          part.isEmpty() ||
          part.equals(".") ||
          part.equals("..") ||
          part.getBytes(StandardCharsets.UTF_8).length > 255
        ) throw invalid();
        spelling += (i == 0 ? "" : "/") + part;
        String key = comparisonKey(spelling);
        boolean directory = i < parts.length - 1;
        if (
          i == 0 &&
          (key.equals("bundle.json") ||
            key.equals("otakit_files.json") ||
            key.equals("otakit-embedded.json"))
        ) throw invalid();
        if (
          spellings.containsKey(key) &&
          (!spellings.get(key).equals(spelling) || !directory || !directories.contains(key))
        ) throw invalid();
        spellings.put(key, spelling);
        if (directory) directories.add(key);
      }
    }
    if (!paths.contains("index.bundle") || !paths.contains("otakit-bundle.json")) throw invalid();
  }

  public void validateDeltaInventory(JSONObject manifest) throws Exception {
    JSONArray files = manifest.getJSONArray("files");
    List<String> paths = new ArrayList<>();
    long total = 0;
    Map<String, Long> sizes = new HashMap<>();
    for (int i = 0; i < files.length(); i++) {
      JSONObject file = files.getJSONObject(i);
      String path = RNManifest.string(file, "path");
      String hash = RNManifest.string(file, "sha256");
      long size = RNManifest.integer(file, "size");
      if (
        !hash.matches("[a-f0-9]{64}") ||
        size < 0 ||
        size > MAX_BYTES - total ||
        (sizes.containsKey(hash) && sizes.get(hash) != size) ||
        ((path.equals("index.bundle") || path.equals("otakit-bundle.json")) && size == 0)
      ) throw invalid();
      paths.add(path);
      sizes.put(hash, size);
      total += size;
    }
    validatePaths(paths);
    if (
      total != RNManifest.integer(manifest, "size") ||
      !inventoryHash(files).equals(manifest.getString("contentHash"))
    ) throw invalid();
  }

  public JSONArray installZip(
    File archive,
    File root,
    JSONObject manifest,
    Map<String, byte[]> keys,
    String rnVersion,
    int bytecodeVersion
  ) throws Exception {
    byte[] bytes = readBounded(archive.toPath(), MAX_BYTES);
    if (
      bytes.length != RNManifest.integer(manifest, "size") ||
      !hash(bytes).equals(manifest.getString("sha256"))
    ) throw invalid();
    if (!manifest.isNull("encryption")) {
      JSONObject e = manifest.getJSONObject("encryption");
      byte[] key = keys.get(e.getString("kid"));
      if (
        key == null ||
        key.length != 32 ||
        !hash(key).substring(0, 16).equals(e.getString("kid")) ||
        !e.getString("alg").equals("AES-256-GCM")
      ) throw invalid();
      byte[] dek = decrypt(
        key,
        Base64.getDecoder().decode(e.getString("wrapNonce")),
        Base64.getDecoder().decode(e.getString("wrappedDek"))
      );
      if (dek.length != 32) throw invalid();
      bytes = decrypt(dek, Base64.getDecoder().decode(e.getString("nonce")), bytes);
    }
    Path temporary = Files.createTempFile(archive.toPath().getParent(), "otakit-plain-", ".zip");
    try {
      Files.write(temporary, bytes);
      try (ZipFile zip = new ZipFile(temporary.toFile())) {
        if (zip.isEncrypted() || zip.isSplitArchive()) throw invalid();
        List<FileHeader> headers = zip.getFileHeaders();
        List<String> paths = new ArrayList<>();
        long total = 0;
        if (headers.size() > 5000) throw invalid();
        for (FileHeader header : headers) {
          byte[] attributes = header.getExternalFileAttributes();
          int type =
            attributes == null || attributes.length != 4
              ? 0
              : (((attributes[3] & 0xff) << 8) | (attributes[2] & 0xff)) & 0170000;
          if (
            header.isDirectory() ||
            header.isEncrypted() ||
            header.getDiskNumberStart() != 0 ||
            (type != 0 && type != 0100000) ||
            (attributes != null && attributes.length == 4 && (attributes[0] & 0x10) != 0) ||
            header.getUncompressedSize() < 0 ||
            header.getUncompressedSize() > MAX_BYTES - total
          ) throw invalid();
          total += header.getUncompressedSize();
          paths.add(header.getFileName());
        }
        validatePaths(paths);
        Path canonical = root.getCanonicalFile().toPath();
        try (java.util.stream.Stream<Path> entries = Files.list(canonical)) {
          if (entries.findAny().isPresent()) throw invalid();
        }
        for (FileHeader header : headers) {
          Path destination = canonical.resolve(header.getFileName());
          if (
            destination.toString().getBytes(StandardCharsets.UTF_8).length + 1 > 4096
          ) throw invalid();
          Files.createDirectories(destination.getParent());
          long written = 0;
          try (
            InputStream input = zip.getInputStream(header);
            OutputStream output = Files.newOutputStream(destination, StandardOpenOption.CREATE_NEW)
          ) {
            byte[] chunk = new byte[8192];
            int length;
            while ((length = input.read(chunk)) != -1) {
              if (length > header.getUncompressedSize() - written) throw invalid();
              output.write(chunk, 0, length);
              written += length;
            }
          }
          if (written != header.getUncompressedSize()) throw invalid();
        }
        return verifyDirectory(root, manifest, rnVersion, bytecodeVersion, paths);
      }
    } finally {
      Files.deleteIfExists(temporary);
    }
  }

  public JSONArray verifyDirectory(
    File directory,
    JSONObject manifest,
    String rnVersion,
    int bytecodeVersion,
    List<String> expectedPaths
  ) throws Exception {
    Path root = directory.getCanonicalFile().toPath();
    if (expectedPaths == null && manifest.has("files")) {
      expectedPaths = new ArrayList<>();
      JSONArray files = manifest.getJSONArray("files");
      for (int i = 0; i < files.length(); i++) expectedPaths.add(
        files.getJSONObject(i).getString("path")
      );
    }
    Map<String, String> original = new HashMap<>();
    if (expectedPaths != null) {
      validatePaths(expectedPaths);
      for (String path : expectedPaths)
        original.put(Normalizer.normalize(path, Normalizer.Form.NFD), path);
    }
    JSONArray files = new JSONArray();
    long total = 0;
    List<String> actualPaths = new ArrayList<>();
    try (java.util.stream.Stream<Path> entries = Files.walk(root)) {
      java.util.Iterator<Path> iterator = entries.iterator();
      while (iterator.hasNext()) {
        Path file = iterator.next();
        BasicFileAttributes attributes = Files.readAttributes(
          file,
          BasicFileAttributes.class,
          LinkOption.NOFOLLOW_LINKS
        );
        if (
          attributes.isSymbolicLink() || (!attributes.isDirectory() && !attributes.isRegularFile())
        ) throw invalid();
        if (attributes.isDirectory()) continue;
        if (
          attributes.size() < 0 || attributes.size() > MAX_BYTES - total || files.length() >= 5000
        ) throw invalid();
        String diskPath = root.relativize(file).toString().replace(File.separatorChar, '/');
        String path =
          expectedPaths == null
            ? diskPath
            : original.get(Normalizer.normalize(diskPath, Normalizer.Form.NFD));
        if (path == null) throw invalid();
        byte[] bytes = readBounded(file, MAX_BYTES - total);
        total += bytes.length;
        files.put(
          new JSONObject().put("path", path).put("sha256", hash(bytes)).put("size", bytes.length)
        );
        actualPaths.add(path);
      }
    }
    validatePaths(actualPaths);
    if (
      (expectedPaths != null && expectedPaths.size() != files.length()) ||
      !inventoryHash(files).equals(manifest.getString("contentHash"))
    ) throw invalid();
    if (manifest.getString("strategy").equals("deltas")) {
      validateDeltaInventory(manifest);
      Map<String, JSONObject> declared = new HashMap<>();
      JSONArray entries = manifest.getJSONArray("files");
      for (int i = 0; i < entries.length(); i++) declared.put(
        entries.getJSONObject(i).getString("path"),
        entries.getJSONObject(i)
      );
      if (total != manifest.getLong("size") || declared.size() != files.length()) throw invalid();
      for (int i = 0; i < files.length(); i++) {
        JSONObject actual = files.getJSONObject(i);
        JSONObject expected = declared.get(actual.getString("path"));
        if (
          expected == null ||
          actual.getLong("size") != expected.getLong("size") ||
          !actual.getString("sha256").equals(expected.getString("sha256"))
        ) throw invalid();
      }
    }
    JSONObject descriptor = new JSONObject(
      new String(
        readBounded(root.resolve("otakit-bundle.json"), 256 * 1024),
        StandardCharsets.UTF_8
      )
    );
    if (
      !RNManifest.string(descriptor, "format").equals("otakit-rn") ||
      RNManifest.integer(descriptor, "formatVersion") != 1 ||
      !RNManifest.string(descriptor, "framework").equals("react-native") ||
      !RNManifest.string(descriptor, "entryPoint").equals("index.bundle") ||
      !RNManifest.string(descriptor, "engine").equals("hermes") ||
      !RNManifest.string(descriptor, "bundleFormat").equals("hermes-bytecode") ||
      !RNManifest.string(descriptor, "reactNativeVersion").equals(rnVersion)
    ) throw invalid();
    for (String field : new String[] { "platform", "runtimeVersion", "version" })
      if (!RNManifest.string(descriptor, field).equals(manifest.getString(field))) throw invalid();
    byte[] bytecode = readBounded(root.resolve("index.bundle"), MAX_BYTES);
    byte[] magic = { (byte) 0xc6, 0x1f, (byte) 0xbc, 0x03, (byte) 0xc1, 0x03, 0x19, 0x1f };
    if (bytecode.length < 12) throw invalid();
    for (int i = 0; i < magic.length; i++) if (bytecode[i] != magic[i]) throw invalid();
    long version = 0;
    for (int i = 0; i < 4; i++) version |= (long) (bytecode[i + 8] & 0xff) << (8 * i);
    if (version != bytecodeVersion) throw invalid();
    if (descriptor.has("expo")) {
      JSONObject expo = descriptor.getJSONObject("expo");
      if (!expo.getString("configFile").equals("expo-config.json")) throw invalid();
      new JSONObject(
        new String(
          readBounded(root.resolve("expo-config.json"), 256 * 1024),
          StandardCharsets.UTF_8
        )
      );
      if (
        expo.has("domRoot") &&
        (!expo.getString("domRoot").equals("www.bundle") ||
          actualPaths
            .stream()
            .noneMatch(path -> path.startsWith("www.bundle/") && path.endsWith(".html")))
      ) throw invalid();
    }
    return files;
  }

  public static byte[] readBounded(Path path, long limit) throws Exception {
    try (
      InputStream input = Files.newInputStream(path);
      ByteArrayOutputStream output = new ByteArrayOutputStream()
    ) {
      byte[] chunk = new byte[8192];
      int length;
      while ((length = input.read(chunk)) != -1) {
        if (length > limit - output.size()) throw invalid();
        output.write(chunk, 0, length);
      }
      return output.toByteArray();
    }
  }

  public static String hash(byte[] bytes) throws Exception {
    StringBuilder result = new StringBuilder();
    for (byte value : MessageDigest.getInstance("SHA-256").digest(bytes))
      result.append(String.format(java.util.Locale.ROOT, "%02x", value & 0xff));
    return result.toString();
  }

  public static String inventoryHash(JSONArray files) throws Exception {
    List<JSONObject> sorted = new ArrayList<>();
    for (int i = 0; i < files.length(); i++) sorted.add(files.getJSONObject(i));
    sorted.sort((left, right) -> {
      byte[] a = left.optString("path").getBytes(StandardCharsets.UTF_8);
      byte[] b = right.optString("path").getBytes(StandardCharsets.UTF_8);
      for (int i = 0; i < Math.min(a.length, b.length); i++) if (
        a[i] != b[i]
      ) return Integer.compare(a[i] & 0xff, b[i] & 0xff);
      return Integer.compare(a.length, b.length);
    });
    StringBuilder canonical = new StringBuilder();
    for (JSONObject file : sorted) {
      if (canonical.length() > 0) canonical.append('\n');
      canonical.append(file.getString("path")).append(':').append(file.getString("sha256"));
    }
    return hash(canonical.toString().getBytes(StandardCharsets.UTF_8));
  }

  private static byte[] decrypt(byte[] key, byte[] nonce, byte[] bytes) throws Exception {
    if (nonce.length != 12 || bytes.length < 16) throw invalid();
    Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
    cipher.init(
      Cipher.DECRYPT_MODE,
      new SecretKeySpec(key, "AES"),
      new GCMParameterSpec(128, nonce)
    );
    return cipher.doFinal(bytes);
  }

  private static Exception invalid() {
    return new Exception("INCOMPATIBLE_ARTIFACT");
  }
}
